src/lib/syntax/lexer.rs

```diff
@@ -466,14 +467,58 @@ impl<'a> Lexer<'a> {
                                         next_ch => buf.push(next_ch),
                                     }
                                 }
-                                TokenData::Comment(buf)
+                                self.push_token(TokenData::Comment(buf));
                             }
-                            '=' => TokenData::Punctuator(Punctuator::AssignDiv),
-                            _ => TokenData::Punctuator(Punctuator::Div),
-                        };
-                        self.push_token(token)
+                            // division, assigndiv or regex literal
+                            _ => {
+                                // if we fail to parse a regex literal, store a copy of the current
+                                // buffer to restore later on
+                                let original_buffer = self.buffer.clone();
+                                // first, try to parse a regex literal
+                                let mut body = String::new();
+                                let mut regex = false;
+                                loop {
+                                    match self.buffer.next() {
+                                        // end of body
+                                        Some('/') => {
+                                            regex = true;
+                                            break;
+                                        }
+                                        // newline/eof not allowed in regex literal
+                                        Some('\n') | Some('\r') | Some('\u{2028}')
+                                        | Some('\u{2029}') | None => break,
+                                        // escape sequence
+                                        Some('\\') => {
+                                            body.push('\\');
+                                            match self.next()? {
+                                                // newline not allowed in regex literal
+                                                '\n' | '\r' | '\u{2028}' | '\u{2029}' => break,
+                                                ch => body.push(ch),
+                                            }
+                                        }
+                                        Some(ch) => body.push(ch),
+                                    }
+                                }
+                                if regex {
+                                    // body was parsed, now look for flags
+                                    let flags = self.take_char_while(char::is_alphabetic)?;
+                                    self.push_token(TokenData::RegularExpression(body, flags));
```

Comment:
I wonder if this should be called "RegularExpressionLiteral" to match up with the other token names?
The specification also refers to them as [RegularExpressionLiteral](https://tc39.es/ecma262/#prod-RegularExpressionLiteral)s too